helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java

```diff
@@ -97,7 +111,15 @@ public CustomRestClientImpl(HttpClient httpClient) {
           kv -> result.put(kv.getKey(), kv.getValue().get(IS_HEALTHY_FIELD).asBoolean()));
       return result;
     };
-    return handleResponse(post(url, payLoads), jsonConverter);
+    try {
+      _semaphore.acquire();
+      return handleResponse(post(url, payLoads), jsonConverter);
+    } catch (InterruptedException e) {
+      LOG.warn("Interrupted while acquiring the semaphore", e);
+    } finally {
+      _semaphore.release();
```

Comment:
That's not the method to lock, right? This is per instance check, it can be parallel without duplication. The most duplicated requests part is partition level checks.

Another thing here, shall we use Semaphore or we can use conditional variable?

Reply:
It is the partition check method. Semaphore allows multiple threads submitting the requests while a single condition can allow up to one(and then it becomes sequential calls), if it's a high volume traffics, performance is not good enough.